src/refda/adm/ietf_dtnma_agent.c

```diff
@@ -2258,6 +2407,170 @@ static void refda_adm_ietf_dtnma_agent_ctrl_ensure_tbr(refda_ctrl_exec_ctx_t *ct
      * |START CUSTOM FUNCTION refda_adm_ietf_dtnma_agent_ctrl_ensure_tbr BODY
      * +-------------------------------------------------------------------------+
      */
+    const cace_ari_t *odm_ns           = refda_ctrl_exec_ctx_get_aparam_index(ctx, 0);
+    const cace_ari_t *ari_obj_name     = refda_ctrl_exec_ctx_get_aparam_index(ctx, 1);
+    const cace_ari_t *ari_obj_enum     = refda_ctrl_exec_ctx_get_aparam_index(ctx, 2);
+    const cace_ari_t *ari_action       = refda_ctrl_exec_ctx_get_aparam_index(ctx, 3);
+    const cace_ari_t *ari_start_time   = refda_ctrl_exec_ctx_get_aparam_index(ctx, 4);
+    const cace_ari_t *ari_period       = refda_ctrl_exec_ctx_get_aparam_index(ctx, 5);
+    const cace_ari_t *ari_max_count    = refda_ctrl_exec_ctx_get_aparam_index(ctx, 6);
+    const cace_ari_t *ari_init_enabled = refda_ctrl_exec_ctx_get_aparam_index(ctx, 7);
+
+    cace_ari_t ari_result;
+
+    refda_agent_t *agent = ctx->runctx->agent;
+
+    REFDA_AGENT_LOCK(agent, );
+    cace_amm_obj_ns_t *odm = cace_amm_obj_store_find_ns(&(agent->objs), odm_ns);
+    REFDA_AGENT_UNLOCK(agent, );
+
+    if (!odm)
+    {
+        CACE_LOG_INFO("ODM not found");
+        return;
+    }
+
+    if (odm->model_id.intenum >= 0)
+    {
+        CACE_LOG_ERR("Invalid model ID, cannot modify an ADM");
+        return;
+    }
+
+    const cace_data_t *obj_name = cace_ari_cget_tstr(ari_obj_name);
+    if (obj_name == NULL || obj_name->ptr == NULL)
+    {
+        CACE_LOG_ERR("Unable to retrieve obj name");
+        return;
+    }
+
+    cace_ari_int obj_id;
+    if (cace_ari_get_int(ari_obj_enum, &obj_id))
+    {
+        CACE_LOG_ERR("Unable to retrieve object ID");
+        return;
+    }
+
+    REFDA_AGENT_LOCK(agent, );
+    bool valid = false;
+    {
+        if (NULL != cace_amm_obj_ns_find_obj_name(odm, CACE_ARI_TYPE_TBR, (const char *)obj_name->ptr))
+        {
+            // FIXME: update fields on existing TBR in this case, instead of just returning??
+            CACE_LOG_INFO("TBR already exists");
+            valid = true;
+            goto ensure_tbr_cleanup;
```

Comment:
Strongly avoid goto! Use the valid state to determine whether to continue normal processing blocks and set the state false here and other uses of goto.

Reply:
Fair enough, refactored to use `valid` and a straightforward control flow for both `ensure` rule operations.